### PR TITLE
Add junit report format

### LINKED ISSUE #404: Junit-Support
**Is your feature request related to a problem? Please describe.**
It's always hard to filter and/or read results on a ci/cd system from json. 
Also visually in a browser.

**Describe the solution you'd like**
JUnit is a widely used format, that can be parsed by a wide range of ci/cd systems including Gitlab + Jenkins. This would help failing builds on results and make them more readable.

**Describe alternatives you've considered**
Parsing Json, converting to tables.

**Additional context**
https://docs.gitlab.com/ee/ci/junit_test_reports.html

cc @zricethezav

regards,
strowi

### PR DESCRIPTION
**Pull request by maltemorgenstern**
### Description:
This merge request adds `junit` as a new report format.

As described in #404 junit reports can be integrated with CI/CD systems as test reports for easier visualisation in the browser and merge requests. I tested these changes with the [GitLab Unit Test Reports](https://docs.gitlab.com/ee/ci/testing/unit_test_reports.html) feature.

That's why the generated `junit` format follows the [GitLab requirements](https://gitlab.com/gitlab-org/gitlab/-/issues/299086). In general it is based on the [IBM junit format](https://www.ibm.com/docs/en/developer-for-zos/14.1.0?topic=formats-junit-xml-format) with some additional and missing attributes (some of them could be added in the future if needed).

Here is an example of the GitLab integration:
<details>
  <summary>Click to expand!</summary>

image

image

</details>

Links/Docs:
- https://www.developer.com/languages/xml-files-golang/
- https://stackoverflow.com/q/26371965
- https://gitlab.com/gitlab-org/gitlab/-/tree/v15.1.2-ee/spec/fixtures/junit
- https://gitlab.com/gitlab-org/gitlab/-/issues/299086
- https://www.ibm.com/docs/en/developer-for-zos/14.1.0?topic=formats-junit-xml-format

This MR closes #404

### Unrelated change

I also added `testdata/expected/report/*.got.*` to the `.gitignore` because they should not be commited to the repo.

### Checklist:

* [x] Does your PR pass tests?
* [x] Have you written new tests for your changes?
* [x] Have you lint your code locally prior to submission?

### CODE REVIEW

**Comment by maltemorgenstern** [comment]
Hey @zricethezav, this PR is ready for review 🚀
Thanks 😃

**Comment by zricethezav** [comment]
@maltemorgenstern would you mind updating this? Also since I do not have any junit experience would you mind keeping an eye on this moving forward when it gets merged in?

**Comment by maltemorgenstern** [comment]
Hey @zricethezav

> Would you mind updating this?

Sure - I will take a look at this the next few days and solve the conflicts!

> Also since I do not have any junit experience would you mind keeping an eye on this moving forward when it gets merged in?

I must admit that I am not really familiar with the `junit` format either - I just looked into it for a bit and created this solution based on other `junit` report implementations/references.

But I should definitely be able to take a look at future PRs and test them with the GitLab integration.

**Comment by maltemorgenstern** [comment]
@zricethezav this PR is now up-to-date again and ready for review/merge.

**Comment by strowi** [comment]
Would really love to see junit support to integrate with CI and have a standardized format. ;)

**Comment by maltemorgenstern** [comment]
Hey @zricethezav,
this pull request is ready for review. If there are any additional checks to be done please let me know!

File: report/junit_test.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+		{
+			testReportName: "empty",
+			expected:       filepath.Join(expectPath, "report", "junit_empty.xml"),
+			findings:       []Finding{},
+		},
+	}
+
+	for _, test := range tests {
+		// create tmp file using os.TempDir()
+		tmpfile, err := os.Create(filepath.Join(tmpPath, test.testReportName+".xml"))
```
Comment (rgmz) [inline report/junit_test.go:68]: @zricethezav `report_test.go` declares `tmpPath` but this file doesn't (re: comment on Discord).

https://github.com/maltemorgenstern/gitleaks/blob/fade1e9ad47e66a233bc4ea1af6d3b46530e4f94/report/report_test.go#L12-L15

Comment (lesaux) [reply report/junit_test.go:68]: can confirm tmpPath is not declared.

Comment (zricethezav) [reply report/junit_test.go:68]: fix't in https://github.com/gitleaks/gitleaks/commit/b1d59bf8905b3dae1b1f84afa76b0e7c6293d59e

**Pull request merged**